src/main/java/com/pnudev/librarysystem/service/BookService.java

```diff
@@ -0,0 +1,141 @@
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+
+        bookMapper.updateBookFromRequestDTO(requestBookDTO, book);
+
+        return bookMapper.toDTO(bookRepository.save(book));
+    }
+
+    public void deleteBookById(Long id) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+
+        if(!book.getBorrowings().isEmpty()) {
+            throw new DeleteFailedException("Book with id: %d is used".formatted(id));
+        }
+
+        bookRepository.deleteById(id);
+    }
+
+    public Page<BookDTO> searchBook(String title, String authorLastName, String categoryName, Pageable pageable) {
+        Specification<Book> specification = Specification.where(null);
+
+        if (title != null && !title.isEmpty()) {
+            specification = specification.and(SpecificationUtils.<Book>fieldContainsIgnoreCase("title", title));
+        }
+        if (authorLastName != null && !authorLastName.isEmpty()){
+            specification = specification.and(
+                    SpecificationUtils.<Book>joinTableFieldContainsIgnoreCase("authors", "lastName", authorLastName)
+            );
+        }
+        if (categoryName != null && !categoryName.isEmpty()){
+            specification = specification.and(
+                    SpecificationUtils.<Book>joinTableFieldContainsIgnoreCase("categories", "name", categoryName)
+            );
+        }
+
+        Page<Book> page = bookRepository.findAll(specification, pageable);
+        return page.map(bookMapper::toDTO);
+    }
+
+    public BookDTO getBookById(Long id) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Book with id: %d not found".formatted(id)));
+        return bookMapper.toDTO(book);
+    }
+
+    public String uploadCoverImage(MultipartFile file) {
+        if(file.isEmpty()) {
+            throw new EmptyFileException("File cannot be empty");
+        }
+
+        String originalFileName = file.getOriginalFilename();
+        String fileExtension = originalFileName.substring(originalFileName.lastIndexOf(".")+1);
+
+        if (!imageTypes.contains(fileExtension)) {
+            throw new FileWrongTypeException(imageTypes);
+        }
+
+        Path uploadPath= Paths.get(uploadDir);
+        if (!Files.exists(uploadPath)) {
+            try {
+                Files.createDirectory(uploadPath);
+            } catch (IOException e) {
+                throw new IOErrorInFileException("Cannot create directory: " + e.getMessage());
+            }
+        }
+
+        String filename = UUID.randomUUID() + "." + fileExtension;
+        Path imagePath = uploadPath.resolve(filename);
+
+        try {
+            Files.copy(file.getInputStream(),imagePath, StandardCopyOption.REPLACE_EXISTING);
```

Comment:
Whitespace (you can do auto-formatting in the intellij, ctrl-alt-L)

Reply:
Formatted